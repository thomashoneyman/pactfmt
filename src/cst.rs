use crate::{
    format::{SourceToken, SpecialForm, Wrapped, FST},
    lexer::Token,
};

#[derive(Debug, PartialEq, Clone)]
pub enum LiteralValue {
    String(String),
    Symbol(String),
    Integer(String),
    Decimal(String),
    Boolean(String),
}

pub type Literal = SourceToken<LiteralValue>;

/// A reference like "my-mod.my-name" or "a.b.c.d"
#[derive(Debug, Clone, PartialEq)]
pub struct Reference {
    pub first: String,
    pub second: String,
    pub rest: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Named {
    Ident(String),
    Reference(Reference),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    /// Simple identifier type like :integer or :time
    Ident(String),
    /// List type like [integer] or [object{schema}]
    List(Box<Type>),
    /// Object type like object{schema} or object{}
    Object(Option<String>),
    /// Schema type like {schema}
    Schema(String),
    /// Module type like module{schema,other.schema}
    Module(Vec<Named>),
}

#[derive(Debug, PartialEq, Clone)]
pub struct IdentifierFields {
    pub identifier: String,
    pub type_annotation: Option<Type>,
}

pub type Identifier = SourceToken<IdentifierFields>;

#[derive(Debug, PartialEq, Clone)]
pub struct App {
    pub left_paren: SourceToken<Token>,
    pub func: Box<Expr>,
    pub args: Vec<Expr>,
    pub right_paren: SourceToken<Token>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct List {
    pub left_bracket: SourceToken<Token>,
    pub members: Vec<Expr>,
    pub right_bracket: SourceToken<Token>,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Identifier(Identifier),
    Literal(Literal),
    Application(App),
    List(List),
}

#[derive(Debug, PartialEq)]
pub struct Arguments {
    pub left_paren: SourceToken<Token>,
    pub args: Vec<Identifier>,
    pub right_paren: SourceToken<Token>,
}

// FIXME: Missing doc/model annotations and bare docstrings. Will probably
// represent each with their own field and typed where the annotation is
// an Option<PrefixSpacing> for doc or a required PrefixSpacing for model,
// and a string for content for a doc and a ...list expr for model(?)
#[derive(Debug, PartialEq)]
pub struct Defun {
    pub left_paren: SourceToken<Token>,
    pub defun: SourceToken<Token>,
    pub name: Identifier,
    pub arguments: Arguments,
    // FIXME: Technically you can't have a list of expressions in the body,
    // it's really just an optional doc annotation, optional model, and then
    // body, which is a single expr.
    pub body: Vec<Expr>,
    pub right_paren: SourceToken<Token>,
}

#[derive(Debug, PartialEq)]
pub enum Toplevel {
    Defun(Defun),
    Expr(Expr),
}

pub fn lower_toplevel(toplevel: Toplevel) -> FST {
    match toplevel {
        Toplevel::Expr(expr) => lower_expr(expr),
        Toplevel::Defun(defun) => lower_defun(defun),
    }
}

fn lower_expr(expr: Expr) -> FST {
    match expr {
        Expr::Literal(lit) => {
            let value = match lit.value {
                LiteralValue::String(s) => s,
                LiteralValue::Symbol(s) => s,
                LiteralValue::Integer(s) => s,
                LiteralValue::Decimal(s) => s,
                LiteralValue::Boolean(s) => s,
            };

            FST::Literal(SourceToken {
                leading: lit.leading,
                value,
                trailing: lit.trailing,
            })
        }

        Expr::Identifier(ident) => {
            let value = if let Some(type_ann) = ident.value.type_annotation {
                format!("{}:{}", ident.value.identifier, format_type(type_ann))
            } else {
                ident.value.identifier
            };

            FST::Literal(SourceToken {
                leading: ident.leading,
                value,
                trailing: ident.trailing,
            })
        }

        Expr::List(list) => FST::List(Wrapped {
            open: SourceToken {
                leading: list.left_bracket.leading,
                value: "[".to_string(),
                trailing: list.left_bracket.trailing,
            },
            inner: list.members.into_iter().map(lower_expr).collect(),
            close: SourceToken {
                leading: list.right_bracket.leading,
                value: "]".to_string(),
                trailing: list.right_bracket.trailing,
            },
        }),

        Expr::Application(app) => FST::SExp(Wrapped {
            open: SourceToken {
                leading: app.left_paren.leading,
                value: "(".to_string(),
                trailing: app.left_paren.trailing,
            },
            inner: std::iter::once(*app.func)
                .chain(app.args)
                .map(lower_expr)
                .collect(),
            close: SourceToken {
                leading: app.right_paren.leading,
                value: ")".to_string(),
                trailing: app.right_paren.trailing,
            },
        }),
    }
}

fn format_type(type_ann: Type) -> String {
    match type_ann {
        Type::Ident(s) => s,
        Type::List(inner) => format!("[{}]", format_type(*inner)),
        Type::Object(schema) => match schema {
            Some(s) => format!("object{{{}}}", s),
            None => "object{}".to_string(),
        },
        Type::Schema(s) => format!("{{{}}}", s),
        Type::Module(names) => {
            let names = names
                .into_iter()
                .map(|n| match n {
                    Named::Ident(s) => s,
                    Named::Reference(r) => {
                        let mut parts = vec![r.first, r.second];
                        parts.extend(r.rest);
                        parts.join(".")
                    }
                })
                .collect::<Vec<_>>()
                .join(",");
            format!("module{{{}}}", names)
        }
    }
}

fn lower_defun(defun: Defun) -> FST {
    FST::SpecialForm(Wrapped {
        open: SourceToken {
            leading: defun.left_paren.leading,
            value: "(".to_string(),
            trailing: defun.left_paren.trailing,
        },
        inner: SpecialForm {
            keyword: SourceToken {
                leading: defun.defun.leading,
                value: "defun".to_string(),
                trailing: defun.defun.trailing,
            },
            sections: vec![
                lower_expr(Expr::Identifier(defun.name)),
                FST::List(Wrapped {
                    open: SourceToken {
                        leading: defun.arguments.left_paren.leading,
                        value: "(".to_string(),
                        trailing: defun.arguments.left_paren.trailing,
                    },
                    inner: defun
                        .arguments
                        .args
                        .into_iter()
                        .map(|arg| lower_expr(Expr::Identifier(arg)))
                        .collect(),
                    close: SourceToken {
                        leading: defun.arguments.right_paren.leading,
                        value: ")".to_string(),
                        trailing: defun.arguments.right_paren.trailing,
                    },
                }),
            ],
            body: defun.body.into_iter().map(lower_expr).collect(),
        },
        close: SourceToken {
            leading: defun.right_paren.leading,
            value: ")".to_string(),
            trailing: defun.right_paren.trailing,
        },
    })
}
