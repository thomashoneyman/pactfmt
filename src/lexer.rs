use logos::Logos;

#[derive(Logos, Debug, PartialEq, Clone)]
pub enum Token {
    #[token("{")]
    LeftBrace,

    #[token("}")]
    RightBrace,

    #[token("(")]
    LeftParen,

    #[token(")")]
    RightParen,

    #[token("[")]
    LeftBracket,

    #[token("]")]
    RightBracket,

    #[token(":=")]
    ColonEquals,

    #[token(":")]
    Colon,

    #[token("::")]
    DoubleColon,

    #[token(",")]
    Comma,

    #[token(".")]
    Dot,

    #[token("true", |_| true)]
    #[token("false", |_| false)]
    Boolean(bool),

    #[token("let*")]
    LetRec,

    #[token("let")]
    Let,

    #[token("if")]
    If,

    #[token("defun")]
    Defun,

    #[token("defcap")]
    DefCap,

    #[token("defconst")]
    DefConst,

    #[token("defschema")]
    DefSchema,

    #[token("deftable")]
    DefTable,

    #[token("defpact")]
    DefPact,

    #[token("interface")]
    Interface,

    #[token("module")]
    Module,

    #[token("bless")]
    Bless,

    #[token("implements")]
    Implements,

    #[token("use")]
    Use,

    #[token("lambda")]
    Lambda,

    #[token("and")]
    And,

    #[token("or")]
    Or,

    #[token("load")]
    Load,

    #[token("@doc")]
    DocAnn,

    #[token("@model")]
    ModelAnn,

    #[token("@event")]
    EventAnn,

    #[token("@managed")]
    ManagedAnn,

    #[token("step-with-rollback")]
    StepWithRollback,

    #[token("enforce")]
    Enforce,

    #[token("enforce-one")]
    EnforceOne,

    #[token("enforce-guard")]
    EnforceGuard,

    #[token("keyset-ref-guard")]
    KeysetRefGuard,

    #[token("step")]
    Step,

    #[token("with-capability")]
    WithCapability,

    #[token("create-user-guard")]
    CreateUserGuard,

    #[token("try")]
    Try,

    #[token("do")]
    Do,

    #[token("suspend")]
    Suspend,

    #[regex(r"-?[0-9]+", |lex| lex.slice().to_string())]
    Integer(String),

    #[regex(r"-?[0-9]+\.[0-9]+", |lex| lex.slice().to_string())]
    Decimal(String),

    // A string which may contain internal escaped quotes like \"
    #[regex(r#""(?:[^"\\]|\\.|\\[\n\r][^"])*""#, |lex| lex.slice().to_string())]
    String(String),

    // A single-quoted string terminated by whitespace, a colon, dot, etc.
    #[regex(r#"'[^\s\t\n\r,:.\"\'\[\]()]+"#, |lex| lex.slice().to_string())]
    Symbol(String),

    // NOTE: This is a vague attempt to capture the many characters allowed in a Pact
    // identifier, though only ASCII, whereas a Pact identifier can also include arbitrary
    // unicode. Will almost certainly have to adjust this in the future.
    #[regex(
        r#"[a-zA-Z%#+\-_&$@<>=?*!|/][a-zA-Z0-9%#+\-_&$@<>=?*!|/]*"#,
        priority = 1,
        callback = |lex| lex.slice().to_string())]
    Ident(String),

    #[regex(r"[ \t]+", logos::skip)]
    Whitespace,

    #[regex(r"(\n|\r\n)+", |lex| lex.slice().to_string())]
    Newlines(String),

    #[regex(r";[^\n]*", |lex| lex.slice().to_string())]
    Comment(String),
}

impl winnow::stream::ContainsToken<Token> for Token {
    #[inline]
    fn contains_token(&self, token: Token) -> bool {
        *self == token
    }
}

impl winnow::stream::ContainsToken<Token> for &'_ [Token] {
    #[inline]
    fn contains_token(&self, token: Token) -> bool {
        self.iter().any(|t| *t == token)
    }
}

impl<const LEN: usize> winnow::stream::ContainsToken<Token> for [Token; LEN] {
    #[inline]
    fn contains_token(&self, token: Token) -> bool {
        self.iter().any(|t| *t == token)
    }
}

impl<const LEN: usize> winnow::stream::ContainsToken<Token> for &'_ [Token; LEN] {
    #[inline]
    fn contains_token(&self, token: Token) -> bool {
        self.iter().any(|t| *t == token)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use logos::Logos;

    fn lex_single(input: &str) -> Token {
        Token::lexer(input).find(Result::is_ok).unwrap().unwrap()
    }

    /// A test macro for single tokens that captures a Rust identifier, Captures a Rust
    /// identifier, expression (input string), and another expression (expected token).
    /// Macro accepts many cases repeated, separated by commas, repeating the indicated
    /// test output for every captured ident/expr set.
    macro_rules! single_token {
        ($($name:ident: $input:expr => $token:expr),*,) => {
            $(
                #[test]
                fn $name() {
                    assert_eq!(lex_single($input), $token);
                }
            )*
        }
    }

    single_token! {
      lex_left_brace: "{" => Token::LeftBrace,
      lex_right_brace: "}" => Token::RightBrace,
      lex_left_paren: "(" => Token::LeftParen,
      lex_right_paren: ")" => Token::RightParen,
      lex_left_bracket: "[" => Token::LeftBracket,
      lex_right_bracket: "]" => Token::RightBracket,
      lex_colon_equals: ":=" => Token::ColonEquals,
      lex_colon: ":" => Token::Colon,
      lex_double_colon: "::" => Token::DoubleColon,
      lex_comma: "," => Token::Comma,
      lex_dot: "." => Token::Dot,
      lex_true: "true" => Token::Boolean(true),
      lex_false: "false" => Token::Boolean(false),
      lex_let_rec: "let*" => Token::LetRec,
      lex_let: "let" => Token::Let,
      lex_if: "if" => Token::If,
      lex_defun: "defun" => Token::Defun,
      lex_defcap: "defcap" => Token::DefCap,
      lex_defconst: "defconst" => Token::DefConst,
      lex_defschema: "defschema" => Token::DefSchema,
      lex_deftable: "deftable" => Token::DefTable,
      lex_defpact: "defpact" => Token::DefPact,
      lex_interface: "interface" => Token::Interface,
      lex_module: "module" => Token::Module,
      lex_bless: "bless" => Token::Bless,
      lex_implements: "implements" => Token::Implements,
      lex_use: "use" => Token::Use,
      lex_lambda: "lambda" => Token::Lambda,
      lex_and: "and" => Token::And,
      lex_or: "or" => Token::Or,
      lex_load: "load" => Token::Load,
      lex_doc_ann: "@doc" => Token::DocAnn,
      lex_model_ann: "@model" => Token::ModelAnn,
      lex_event_ann: "@event" => Token::EventAnn,
      lex_managed_ann: "@managed" => Token::ManagedAnn,
      lex_step_with_rollback: "step-with-rollback" => Token::StepWithRollback,
      lex_enforce: "enforce" => Token::Enforce,
      lex_enforce_one: "enforce-one" => Token::EnforceOne,
      lex_enforce_guard: "enforce-guard" => Token::EnforceGuard,
      lex_keyset_ref_guard: "keyset-ref-guard" => Token::KeysetRefGuard,
      lex_step: "step" => Token::Step,
      lex_with_capability: "with-capability" => Token::WithCapability,
      lex_create_user_guard: "create-user-guard" => Token::CreateUserGuard,
      lex_try: "try" => Token::Try,
      lex_do: "do" => Token::Do,
      lex_suspend: "suspend" => Token::Suspend,

      // These entries use regex instead of full string matches
      lex_integer: "42" => Token::Integer("42".into()),
      lex_neg_integer: "-42" => Token::Integer("-42".into()),
      lex_decimal: "3.14" => Token::Decimal("3.14".into()),
      lex_neg_decimal: "-3.14" => Token::Decimal("-3.14".into()),
      lex_ident: "add" => Token::Ident("add".into()),
      lex_complex_ident: "add-two!" => Token::Ident("add-two!".into()),
      lex_string: "\"hello\"" => Token::String("\"hello\"".into()),
      lex_escaped_string: "\"he\\\"llo\"" => Token::String("\"he\\\"llo\"".into()),
      lex_symbol: "'mysym" => Token::Symbol("'mysym".into()),
      lex_comment: "; comment" => Token::Comment("; comment".into()),
      lex_unix_newline: "\n" => Token::Newlines("\n".into()),
      lex_windows_newline: "\r\n" => Token::Newlines("\r\n".into()),
    }

    #[test]
    fn test_whitespace_sequence() {
        let input = "  \n  \t\t";
        let mut lexer = Token::lexer(input);
        while let Some(token_result) = lexer.next() {
            println!(
                "Token result: {:?} at position {:?}",
                token_result,
                lexer.span()
            );
        }

        // Keep the original assertion for when we fix the issue
        let tokens: Vec<_> = Token::lexer(input)
            .map(|result| result.expect("Failed to lex token"))
            .collect();
        assert_eq!(tokens, vec![Token::Newlines("\n".into())]);
    }
}
