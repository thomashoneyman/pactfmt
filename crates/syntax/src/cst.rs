use crate::lexer::Token;

#[derive(Debug, PartialEq, Clone)]
pub enum Spacing {
    Newline(usize),
    Comment(String),
}

#[derive(Debug, PartialEq, Clone)]
pub struct SourceToken<T> {
    pub leading: Vec<Spacing>,
    pub value: T,
    pub trailing: Vec<Spacing>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Wrapped<T> {
    pub open: SourceToken<String>,
    pub inner: T,
    pub close: SourceToken<String>,
}

#[derive(Debug, PartialEq, Clone)]
pub enum LiteralValue {
    String(String),
    Symbol(String),
    Integer(String),
    Decimal(String),
    Boolean(String),
}

pub type Literal = SourceToken<LiteralValue>;

/// A reference like "my-mod.my-name" or "a.b.c.d"
#[derive(Debug, Clone, PartialEq)]
pub struct Reference {
    pub first: String,
    pub second: String,
    pub rest: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Named {
    Ident(String),
    Reference(Reference),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    /// Simple identifier type like :integer or :time
    Ident(String),
    /// List type like [integer] or [object{schema}]
    List(Box<Type>),
    /// Object type like object{schema} or object{}
    Object(Option<String>),
    /// Schema type like {schema}
    Schema(String),
    /// Module type like module{schema,other.schema}
    Module(Vec<Named>),
}

#[derive(Debug, PartialEq, Clone)]
pub struct IdentifierFields {
    pub identifier: String,
    pub type_annotation: Option<Type>,
}

pub type Identifier = SourceToken<IdentifierFields>;

#[derive(Debug, PartialEq, Clone)]
pub struct App {
    pub left_paren: SourceToken<Token>,
    pub func: Box<Expr>,
    pub args: Vec<Expr>,
    pub right_paren: SourceToken<Token>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct List {
    pub left_bracket: SourceToken<Token>,
    pub members: Vec<(Expr, Option<SourceToken<Token>>)>,
    pub right_bracket: SourceToken<Token>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Object {
    pub left_brace: SourceToken<Token>,
    pub members: Vec<(SourceToken<String>, SourceToken<Token>, Expr)>,
    pub right_brace: SourceToken<Token>,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Identifier(Identifier),
    Literal(Literal),
    Application(App),
    List(List),
    Object(Object),
}

#[derive(Debug, PartialEq)]
pub struct Arguments {
    pub left_paren: SourceToken<Token>,
    pub args: Vec<Identifier>,
    pub right_paren: SourceToken<Token>,
}

// FIXME: Missing doc/model annotations and bare docstrings. Will probably
// represent each with their own field and typed where the annotation is
// an Option<PrefixSpacing> for doc or a required PrefixSpacing for model,
// and a string for content for a doc and a ...list expr for model(?)
#[derive(Debug, PartialEq)]
pub struct Defun {
    pub left_paren: SourceToken<Token>,
    pub defun: SourceToken<Token>,
    pub name: Identifier,
    pub arguments: Arguments,
    // FIXME: Technically you can't have a list of expressions in the body,
    // it's really just an optional doc annotation, optional model, and then
    // body, which is a single expr.
    pub body: Vec<Expr>,
    pub right_paren: SourceToken<Token>,
}

#[derive(Debug, PartialEq)]
pub enum Toplevel {
    Defun(Defun),
    Expr(Expr),
}
