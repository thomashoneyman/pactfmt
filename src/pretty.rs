use pretty::RcDoc;

use crate::cst::*;
use crate::lexer::Token;

pub trait Pretty {
    fn pretty(&self) -> RcDoc<()>;
}

fn format_spacing(spacing: &[Spacing]) -> RcDoc<()> {
    if spacing.is_empty() {
        return RcDoc::nil();
    }

    let docs: Vec<RcDoc<()>> = spacing
        .iter()
        .map(|s| match s {
            Spacing::NewlineOne => RcDoc::hardline(),
            Spacing::NewlineMany => RcDoc::hardline().append(RcDoc::hardline()),
            Spacing::Comment(text) => RcDoc::text(text),
        })
        .collect();

    RcDoc::concat(docs)
}

fn has_newline(spacing: &[Spacing]) -> bool {
    spacing
        .iter()
        .any(|s| matches!(s, Spacing::NewlineOne | Spacing::NewlineMany))
}

impl Pretty for Token {
    fn pretty(&self) -> RcDoc<()> {
        match *self {
            Token::LeftBrace => RcDoc::text("{"),
            Token::RightBrace => RcDoc::text("}"),
            Token::LeftParen => RcDoc::text("("),
            Token::RightParen => RcDoc::text(")"),
            Token::LeftBracket => RcDoc::text("["),
            Token::RightBracket => RcDoc::text("]"),
            Token::ColonEquals => RcDoc::text(":="),
            Token::Colon => RcDoc::text(":"),
            Token::DoubleColon => RcDoc::text("::"),
            Token::Comma => RcDoc::text(","),
            Token::Dot => RcDoc::text("."),
            Token::Boolean(bool) => match bool {
                true => RcDoc::text("true"),
                false => RcDoc::text("false"),
            },
            Token::LetRec => RcDoc::text("let*"),
            Token::Let => RcDoc::text("let"),
            Token::If => RcDoc::text("if"),
            Token::Defun => RcDoc::text("defun"),
            Token::DefCap => RcDoc::text("defcap"),
            Token::DefConst => RcDoc::text("defconst"),
            Token::DefSchema => RcDoc::text("defschema"),
            Token::DefTable => RcDoc::text("deftable"),
            Token::DefPact => RcDoc::text("defpact"),
            Token::Interface => RcDoc::text("interface"),
            Token::Module => RcDoc::text("module"),
            Token::Bless => RcDoc::text("bless"),
            Token::Implements => RcDoc::text("implements"),
            Token::Use => RcDoc::text("use"),
            Token::Lambda => RcDoc::text("lambda"),
            Token::And => RcDoc::text("and"),
            Token::Or => RcDoc::text("or"),
            Token::Load => RcDoc::text("load"),
            Token::DocAnn => RcDoc::text("@doc"),
            Token::ModelAnn => RcDoc::text("@model"),
            Token::EventAnn => RcDoc::text("@event"),
            Token::ManagedAnn => RcDoc::text("@managed"),
            Token::Step => RcDoc::text("step"),
            Token::StepWithRollback => RcDoc::text("step-with-rollback"),
            Token::Enforce => RcDoc::text("enforce"),
            Token::EnforceOne => RcDoc::text("enforce-one"),
            Token::WithCapability => RcDoc::text("with-capability"),
            Token::CreateUserGuard => RcDoc::text("create-user-guard"),
            Token::Try => RcDoc::text("try"),
            Token::Do => RcDoc::text("do"),
            Token::Suspend => RcDoc::text("suspend"),
            Token::Integer(ref str) => RcDoc::text(str),
            Token::Decimal(ref str) => RcDoc::text(str),
            Token::String(ref str) => RcDoc::text(str),
            Token::Symbol(ref str) => RcDoc::text(str),
            Token::Ident(ref str) => RcDoc::text(str),
            Token::Whitespace => {
                unreachable!("invariant violated: whitespace was consumed as arbitrary token")
            }
            Token::Newlines(_) => {
                unreachable!("invariant violated: newline was consumed as arbitrary token")
            }
            Token::Comment(_) => {
                unreachable!("invariant violated: comment was consumed as arbitrary token")
            }
        }
    }
}

impl<T> Pretty for Positioned<T>
where
    T: Pretty,
{
    fn pretty(&self) -> RcDoc<()> {
        let leading: RcDoc<()> = format_spacing(&self.0);
        leading.append(self.1.pretty())
    }
}

impl Pretty for Type {
    fn pretty(&self) -> RcDoc<()> {
        match &self {
            Type::Ident(s) => RcDoc::text(s),
            Type::List(ty) => RcDoc::text("[")
                .append(ty.pretty())
                .append(RcDoc::text("]")),
            Type::Object(ty) => RcDoc::text("object{")
                .append(RcDoc::text(ty.as_deref().unwrap_or("")))
                .append(RcDoc::text("}")),
            Type::Schema(ty) => RcDoc::text("{")
                .append(RcDoc::text(ty))
                .append(RcDoc::text("}")),
            Type::Module(parts) => {
                let mut doc = RcDoc::text("module{");
                let mut first = true;
                for part in parts {
                    if !first {
                        doc = doc.append(RcDoc::text(","));
                    }
                    doc = doc.append(part.pretty());
                    first = false;
                }
                doc.append(RcDoc::text("}"))
            }
        }
    }
}

impl Pretty for IdentifierFields {
    fn pretty(&self) -> RcDoc<()> {
        match self.type_annotation {
            None => RcDoc::text(&self.identifier),
            Some(ref ty) => RcDoc::text(&self.identifier)
                .append(RcDoc::text(":"))
                .append(ty.pretty()),
        }
    }
}

impl Pretty for LiteralValue {
    fn pretty(&self) -> RcDoc<()> {
        match self {
            LiteralValue::String(str) => RcDoc::text(str),
            LiteralValue::Symbol(str) => RcDoc::text(str),
            LiteralValue::Integer(str) => RcDoc::text(str),
            LiteralValue::Decimal(str) => RcDoc::text(str),
            LiteralValue::Boolean(str) => RcDoc::text(str),
        }
    }
}

impl Pretty for App {
    fn pretty(&self) -> RcDoc<()> {
        let l_paren = format_spacing(&self.left_paren).append("(");
        let r_paren = format_spacing(&self.right_paren).append(")");

        // If any argument has a newline, we'll use multi-line format
        let multiline = self.args.iter().any(|arg| match arg {
            Expr::Identifier((spacing, _))
            | Expr::Literal((spacing, _))
            | Expr::Application(App {
                left_paren: spacing,
                ..
            }) => has_newline(spacing),
        });

        if multiline {
            // Each argument gets a newline before it, unless it already has one
            let args_doc = RcDoc::intersperse(
                self.args.iter().map(|arg| match arg {
                    Expr::Identifier((spacing, _))
                    | Expr::Literal((spacing, _))
                    | Expr::Application(App {
                        left_paren: spacing,
                        ..
                    }) if !has_newline(spacing) => RcDoc::hardline().append(arg.pretty()),
                    _ => arg.pretty(),
                }),
                RcDoc::nil(),
            )
            .nest(2);

            RcDoc::hardline()
                .append(l_paren)
                .append(self.func.pretty())
                .append(args_doc)
                .append(r_paren)
        } else {
            let args = match self.args.split_first() {
                None => RcDoc::nil(),
                Some((first, rest)) => {
                    let mut doc = RcDoc::space().append(first.pretty());
                    for arg in rest {
                        doc = doc.append(RcDoc::space()).append(arg.pretty());
                    }
                    doc
                }
            };

            l_paren
                .append(self.func.pretty())
                .append(args)
                .append(r_paren)
        }
    }
}

impl Pretty for Expr {
    fn pretty(&self) -> RcDoc<()> {
        match self {
            Expr::Identifier(id) => id.pretty(),
            Expr::Literal(lit) => lit.pretty(),
            Expr::Application(app) => app.pretty(),
        }
    }
}

impl Pretty for Arguments {
    fn pretty(&self) -> RcDoc<()> {
        let l_paren = format_spacing(&self.left_paren).append("(");
        let r_paren = format_spacing(&self.right_paren).append(")");

        // Check if any argument has a newline
        let multiline = self.args.iter().any(|arg| match arg {
            (spacing, _) => has_newline(spacing),
        });

        if multiline {
            match self.args.split_first() {
                Some((first, rest)) => l_paren
                    .append(RcDoc::space())
                    .append(first.pretty())
                    .append(
                        RcDoc::concat(rest.iter().map(|arg| match arg {
                            (spacing, _) if !has_newline(spacing) => {
                                RcDoc::hardline().append(arg.pretty())
                            }
                            _ => arg.pretty(),
                        }))
                        .nest(2),
                    )
                    .append(RcDoc::hardline())
                    .append(r_paren)
                    .nest(2),
                None => l_paren.append(r_paren),
            }
        } else {
            l_paren
                .append(RcDoc::concat(self.args.iter().map(Pretty::pretty)))
                .append(r_paren)
        }
    }
}

impl Pretty for Defun {
    fn pretty(&self) -> RcDoc<()> {
        let l_paren = format_spacing(&self.left_paren).append("(");
        let defun = format_spacing(&self.defun)
            .append("defun")
            .append(RcDoc::space());
        let name = format_spacing(&self.name.0)
            .append(self.name.1.pretty())
            .append(RcDoc::space());
        let r_paren = format_spacing(&self.right_paren)
            .append(")")
            .append(RcDoc::space());

        let body = RcDoc::concat(self.body.iter().map(|expr| match expr {
            Expr::Identifier((spacing, _))
            | Expr::Literal((spacing, _))
            | Expr::Application(App {
                left_paren: spacing,
                ..
            }) if !has_newline(spacing) => RcDoc::hardline().append(expr.pretty()),
            _ => expr.pretty(),
        }));

        l_paren
            .append(defun)
            .append(name)
            .append(self.arguments.pretty())
            .append(body.nest(2))
            .append(r_paren)
    }
}

impl Pretty for Named {
    fn pretty(&self) -> RcDoc<()> {
        match self {
            Named::Ident(s) => RcDoc::text(s),
            Named::Reference(r) => r.pretty(),
        }
    }
}

impl Pretty for Reference {
    fn pretty(&self) -> RcDoc<()> {
        let mut doc = RcDoc::text(&self.first)
            .append(RcDoc::text("."))
            .append(RcDoc::text(&self.second));

        for part in &self.rest {
            doc = doc.append(RcDoc::text(".")).append(RcDoc::text(part));
        }
        doc
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::lexer::*;
    use crate::parser;
    use crate::pretty::Pretty;

    use logos::Logos;
    use test_each_file::test_each_file;

    fn lex(input: &str) -> Vec<Token> {
        Token::lexer(input).filter_map(|token| token.ok()).collect()
    }

    test_each_file! { in "./fixtures" => |content: &str| {
        let tokens = lex(content);
        let mut input = tokens.as_slice();
        let parsed = parser::defun(&mut input).expect("parsing");

        let doc = parsed.pretty();
        let mut buffer = String::new();
        doc.render_fmt(80, &mut buffer).expect("pretty printing");

        insta::assert_snapshot!(buffer);
    }}

    #[test]
    fn test_format_spacing() {
        let spacing = vec![
            Spacing::NewlineMany,
            Spacing::Comment("; hello".to_string()),
            Spacing::NewlineOne,
        ];

        let doc = format_spacing(&spacing);
        let mut output = String::new();
        doc.render_fmt(80, &mut output).unwrap();

        assert_eq!(output, "\n\n; hello\n");
    }
}
