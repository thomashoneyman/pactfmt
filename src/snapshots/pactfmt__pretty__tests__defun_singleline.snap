---
source: src/pretty.rs
expression: buffer
---
(defun square (x) (+ x x))
